proxy/src/control/discovery.rs

```diff
@@ -372,23 +396,57 @@ impl<T> DestinationSet<T>
     where T: HttpService<RequestBody = BoxBody, ResponseBody = RecvBody>,
           T::Error: fmt::Debug
 {
-    fn poll_destination(&mut self, auth: &DnsNameAndPort, mut rx: UpdateRx<T>)
-                        -> DestinationServiceQuery<T>
+    fn reset_dns_query(
+        &mut self,
+        dns_resolver: &dns::Resolver,
+        delay: Duration,
+        authority: &DnsNameAndPort)
+    {
+        trace!("resetting DNS query for {} with delay {:?}", authority.host, delay);
+        self.reset_on_next_modification();
+        self.dns_query = Some(dns_resolver.resolve_all_ips(delay, &authority.host));
+    }
+
+    fn poll_destination_service(
+        &mut self,
+        dns_resolver: &dns::Resolver,
```

Comment:
it's mildly surprising that this takes a `dns_resolver` (as I'd expect this to... only poll the destination service).  It seems this is necessary so that it can initialize a dns query, which is polled in the followup call to `poll_dns.`  I'm fine with this as it is, but it seems like something it would be nice to separate further in a future change.